rdb/db.go

```diff
@@ -0,0 +1,245 @@
+//go:build rocksdb
+// +build rocksdb
+
+package rdb
+
+// #include <stdlib.h>
+// #include "rocksdb/c.h"
+import "C"
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"runtime"
+	"unsafe"
+
+	tmdb "github.com/line/tm-db/v2"
+	"github.com/line/tm-db/v2/internal/util"
+)
+
+type RDB struct {
+	name   string
+	fn     string
+	db     *C.rocksdb_t
+	opts   *C.rocksdb_options_t
+	ropts  *C.rocksdb_readoptions_t
+	wopts  *C.rocksdb_writeoptions_t
+	wsopts *C.rocksdb_writeoptions_t
+	wlpopts *C.rocksdb_writeoptions_t
+}
+
+type rdbIterator struct {
+	db         *RDB
+	it         *C.rocksdb_iterator_t
+	ropts      *C.rocksdb_readoptions_t
+	reverse    bool
+	lowerBound []byte
+	upperBound []byte
+}
+
+func cerror(cerr *C.char) error {
+	if cerr == nil {
+		return nil
+	}
+	err := errors.New(C.GoString(cerr))
+	C.free(unsafe.Pointer(cerr))
+	return err
+}
+
+func b2c(b []byte) *C.char {
+	if len(b) == 0 {
+		return nil
+	}
+	return (*C.char)(unsafe.Pointer(&b[0]))
+}
+
+func NewDB(name string, dir string) (*RDB, error) {
+	var cerr *C.char
+
+	fn := filepath.Join(dir, name+".db")
+
+	bbto := C.rocksdb_block_based_options_create()
+	C.rocksdb_block_based_options_set_block_cache(bbto, C.rocksdb_cache_create_lru(C.size_t(1<<30)))
+	C.rocksdb_block_based_options_set_filter_policy(bbto, C.rocksdb_filterpolicy_create_bloom(C.int(10)))
+	defer C.rocksdb_block_based_options_destroy(bbto)
+
+	opts := C.rocksdb_options_create()
+	C.rocksdb_options_set_block_based_table_factory(opts, bbto)
+	C.rocksdb_options_set_create_if_missing(opts, 1)
+	C.rocksdb_options_increase_parallelism(opts, C.int(runtime.NumCPU()))
+	C.rocksdb_options_optimize_level_style_compaction(opts, 512*1024*1024)
+	C.rocksdb_options_optimize_level_style_compaction(opts, 512*1024*1024)
+	C.rocksdb_options_set_enable_pipelined_write(opts, 1)
+
+	ropts := C.rocksdb_readoptions_create()
+	wopts := C.rocksdb_writeoptions_create()
+	wsopts := C.rocksdb_writeoptions_create()
+	C.rocksdb_writeoptions_set_sync(wsopts, C.uchar(1))
+	wlpopts := C.rocksdb_writeoptions_create()
+	C.rocksdb_writeoptions_set_low_pri(wlpopts, C.uchar(1))
+
+	db := C.rocksdb_open(opts, b2c([]byte(fn)), &cerr)
+	if cerr != nil {
+		C.rocksdb_options_destroy(opts)
+		C.rocksdb_writeoptions_destroy(wopts)
+		C.rocksdb_writeoptions_destroy(wsopts)
+		C.rocksdb_writeoptions_destroy(wlpopts)
+		return nil, cerror(cerr)
+	}
+	return &RDB{
+		name:   name,
+		fn:     fn,
+		db:     db,
+		opts:   opts,
+		ropts:  ropts,
+		wopts:  wopts,
+		wsopts: wsopts,
+		wlpopts: wlpopts,
+	}, nil
+}
+
+func (db *RDB) Name() string {
+	return db.name
+}
+
+func (db *RDB) Get(key []byte) ([]byte, error) {
+	var cerr *C.char
+	var cvl C.size_t
+	ck := b2c(key)
+	cv := C.rocksdb_get(db.db, db.ropts, ck, C.size_t(len(key)), &cvl, &cerr)
+	if cerr != nil {
+		err := cerror(cerr)
```

Comment:
I think to need to free `unsafe.Pointer(cv)` before return. How about using `defer` of golang?

Reply:
I think cv is nil if rocksdb_get return an error. So, maybe free(unsafe.Pointer(cv)) isn't needed in this case.

Reply:
I think cv is nil if rocksdb_get return an error. So, maybe free(unsafe.Pointer(cv)) isn't needed in this case.